### ISSUE TITLE
Virtual Filesystem moves libraries to incompatible paths

### ISSUE THREAD

**Issue by novafacing** [trailofbits/maat#54]
As discussed in #49, the issue of libraries not being recursively loaded has (I think!) been fixed, but there is still the problem of libraries being placed in the virtual filesystem at paths different from their original location. For example using the same setup as in #49 with AIS-Lite and modifying the code a bit:

```python
from collections import defaultdict
from typing import Dict, Tuple
from pathlib import Path
from argparse import ArgumentParser
from typing import Dict, List, Optional
from maat import MaatEngine, OS, ARCH, BIN, EVENT, WHEN
from angr import Project
from bisect import insort


class Stage1Maat:
    """Test using maat to run stage 1 instead of triton..."""

    BASE = 0x400000

    def __init__(
        self,
        binary: Path,
        input_file: Path,
        args: Optional[List[str]] = None,
        libdirs: Optional[List[str]] = None,
    ) -> None:
        """
        Set up the maat engine with our binary.

        :param binary: Path to the binary to run
        :param args: Arguments to pass to the binary
        """
        self.binary = binary
        self.input_file = input_file
        self.args = args
        self.libdirs = (
            list(map(lambda l: str(l.resolve()), libdirs))
            if libdirs is not None
            else []
        )
        self.project = Project(
            str(self.binary.resolve()),
            load_options={
                "auto_load_libs": True,
                "ld_path": self.libdirs,
                "use_system_libs": False,
                "skip_libs": ["libc.so.6", "libm.so.6"],
            },
        )
        self.cfg = self.project.analyses.CFGFast(normalize=True)
        self.engine = MaatEngine(ARCH.X64, OS.LINUX)
        assert binary.is_file(), f"{binary} is not a file or doesn't exist."

        print(f"Loading binary {self.binary} with libdirs {self.libdirs}")

        self.engine.load(
            str(binary.resolve()),
            BIN.ELF64,
            args=args if args is not None else [],
            libdirs=self.libdirs,
        )

        self.set_callbacks()
        self.maps = {}

    def run(self) -> None:
        """
        Run the binary.
        """
        self.engine.run()

    def get_mappings(self, engine: MaatEngine) -> Dict[str, List[Tuple[int]]]:
        """
        Get the mappings of the binary.

        :param engine: Maat engine
        """
        mappings = defaultdict(list)
        rawmaps = list(map(lambda l: l.strip(), str(engine.mem).splitlines()))
        maps = map(
            lambda l: l.split(),
            filter(
                lambda l: l and l.startswith("0x"),
                rawmaps[: rawmaps.index("Page permissions:")],
            ),
        )

        for mp in maps:
            insort(mappings[mp[2]], (int(mp[0], 16), int(mp[1], 16)))

        return dict(mappings)

    def loader_callback(self, engine: MaatEngine) -> None:
        """
        Callback when the loader is running.

        :param engine: Maat engine
        """
        maps = self.get_mappings(engine)
        if maps != self.maps:
            print(f"Maps changed: {maps}")
            self.maps = maps

    def got_to_main(self, _: MaatEngine) -> None:
        """
        Callback when we get to main

        :param engine: Maat engine
        """
        print("Got to main!")

    def set_callbacks(self) -> None:
        """
        Set a callback to...set the rest of our callbacks (lol) once the program
        is loaded.
        """
        main_addr = self.project.loader.find_symbol("main").rebased_addr
        print(f"Setting callback at {main_addr:#016x}")
        maps = self.get_mappings(self.engine)
        print(f"Initial maps {maps}")
        self.engine.hooks.add(
            EVENT.EXEC,
            WHEN.BEFORE,
            filter=(
                maps["ld-linux-x86-64.so.2"][0][0],
                maps["ld-linux-x86-64.so.2"][-1][1],
            ),
            callbacks=[self.loader_callback],
            group="loader",
        )

        self.engine.hooks.add(
            EVENT.EXEC,
            WHEN.BEFORE,
            filter=main_addr,
            callbacks=[self.got_to_main],
            group="setup",
        )


if __name__ == "__main__":
    parser = ArgumentParser(prog="repro")
    parser.add_argument(
        "--binary", type=Path, required=True, help="Path to binary to execute."
    )
    parser.add_argument(
        "--input", type=Path, required=True, help="Path to input file to read."
    )
    parser.add_argument("--args", nargs="*", help="Arguments to pass to binary.")
    parser.add_argument(
        "--libdirs", nargs="*", type=Path, help="Library directories to load."
    )
    cli_args = parser.parse_args()

    repro = Stage1Maat(
        cli_args.binary,
        cli_args.input,
        cli_args.args,
        cli_args.libdirs + [Path("/lib/x86_64-linux-gnu"), Path("/lib64")],
    )
    repro.run()
```

We end up getting:
```
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 139263), (139264, 172031), (176128, 188415)], 'Interp.': [(188416, 4382719)], 'Heap': [(4382720, 8417279)], 'map_anon': [(67108864, 67117055)], '/usr/lib/libcgc.so': [(67117056, 67125247), (67125248, 67133439), (67133440, 67137535), (67137536, 67145727)], 'Stack': [(8796090925056, 8796093022207)]}
AIS-Lite: /usr/lib/libcgc.so: no version information available (required by AIS-Lite)
AIS-Lite: /usr/lib/libcgc.so: no version information available (required by /usr/lib/libcgc.so)
AIS-Lite: symbol lookup error: AIS-Lite: undefined symbol: __libc_start_main, version GLIBC_2.2.5
[Error] Exception in CALLOTHER handler: SYSCALL: EnvEmulator: syscall '231' not supported for emulation
[Error] Unexpected error when processing IR instruction, aborting...
```

Syscall 231 is exit_group, and we're exiting partway through loading, but the thing that stands out to me is that `libcgc.so` is tossed into `/usr/lib` instead of being emulated in the same directory it is on the host machine. Of course we can solve this with patchelf, but is there another way to either use the same `virtual_path` argument passed to `MaatEngine.load` to also specify virtual paths for libraries?

I've repackaged the CGC binaries, so the `.tar.gz` attached to #49 can be compiled on a tester's machine with `cmake -DCMAKE_EXE_LINKER_FLAGS='-no-pie -fno-pie' . && make` to duplicate the configuration I have on my host machine.

**Comment by novafacing** [comment]
If you're accepting PRs, I can also try and implement this using the same logic as `virtual_path` uses :)

**Comment by Boyan-MILANOV** [comment]
> Of course we can solve this with patchelf, but is there another way to either use the same virtual_path argument passed to MaatEngine.load to also specify virtual paths for libraries?

I think the best solution would be to dump the `virtual_path` argument in favor of a `virtual_fs` argument which would consist in a list of arbitrary file locations in the emulated filesystem, e.g `virtual_fs=["/home/me/AIS-Lite", "/reface/evaluation/cgc/AIS-Lite/lib_include/libcgc.so", ...]`. When adding a library file we can check if a specific path was specified for it and add it accordingly, similarly to what `virtual_path` does. If the library isn't present in `virtual_fs` then we keep the current default of `/usr/lib` (and `/` for the binary)

> If you're accepting PRs, I can also try and implement this using the same logic as virtual_path uses :)

We'd be more than happy if you came up with a PR for this 😄 . To give you a few pointers:

First modify the prototypes in https://github.com/trailofbits/maat/blob/master/src/include/maat/loader.hpp to replace `const std::string& virtual_path` with `const std::list<std::string>& virtual_fs`. Do it also in [their .cpp files](https://github.com/trailofbits/maat/tree/master/src/loader)

Then you can pass `virtual_fs` to this function and substitute explicit paths to `/usr/lib` when appropriate https://github.com/trailofbits/maat/blob/c98910985e41770d15a4f4066afdcb8127ed9d50/src/loader/loader_lief_elf.cpp#L574-L614 

Last, update the python bindings for the `MaatEngine.load()` method: https://github.com/trailofbits/maat/blob/c98910985e41770d15a4f4066afdcb8127ed9d50/bindings/python/py_engine.cpp#L91
To convert `virtual_fs` from a list of python strings to a list of C++ strings you can copy what we do for the `libdirs` and `ignore_libs` arguments: https://github.com/trailofbits/maat/blob/c98910985e41770d15a4f4066afdcb8127ed9d50/bindings/python/py_engine.cpp#L172-L189

Don't hesitate to open a draft PR and to reach out to us with more questions!

**Comment by novafacing** [comment]
This is partially implemented on my fork: https://github.com/novafacing/maat/tree/feat/virtual_fs just need to fix a couple remaining issues. Example of `load` now looks like:

```python
vfs_arg = {'libcgc.so': '/AIS-Lite/lib_include/', 'AIS-Lite': '/AIS-Lite', 'libtiny-AES128-C.so': AIS-Lite/lib_include/tiny-AES128-C/'}
self.engine.load(
    str(binary.resolve()),
    BIN.ELF64,
    args=args if args is not None else [],
    libdirs=self.libdirs,
    virtual_fs=vfs_arg,
)
```

Homework interfering with finishing it in a timely manner but it's getting there :)

**Comment by Boyan-MILANOV** [comment]
Awesome! It looks great so far. You're getting that `Failed to add the binary in the virtual filesystem [...]` error because of this code: 

https://github.com/novafacing/maat/blob/43ae9209866848b124a696a6ec98967731f6cf9c/src/loader/loader_lief.cpp#L40-L47

In the `virtual_fs` parameter you're passing the directory of the binary (`'AIS-Lite': '/AIS-Lite/'`) while the code expects the directory followed by the filename (`/AIS-Lite/AIS-Lite`). I'm not sure whether we want users to specify the paths with or without the filenames. On one hand it allows to change the name of the files in the virtual FS if needed, on the other hand it can be redundant to rewrite the filenames if we want to keep them the same (`'libABC.so': '/a/b/c/libABC.so'` while `'libABC.so':'/a/b/c/'` would suffice). 

In my opinion the best solution is to offer both options. If a path in `virtual_fs` ends with a path separator (`'/'`) then we append the library name to it, otherwise we consider that the last element of the path is the filename on the virtual FS: 
```
virtual_fs = {
    "libABC.so":"/a/b/c/" # Puts libABC.so in directory /a/b/c/
    "libDEF.so":"/d/e/f" # Renames libDEF.so as 'f' and put it in directory /d/e/
}
```

As regards implementation, there's a `std::string path_separator` private field of `FileSystem` for which we can add a public getter: https://github.com/trailofbits/maat/blob/master/src/include/maat/env/filesystem.hpp#L222
Then its just a matter of checking if paths in `virtual_fs` end with the `path_separator`:

- if they end with `path_separator`: append the filename to it
- if they don't: consider the path as the full path including the filename, and don't modify it

@novafacing: do you think adding this is feasible? You can always open a draft PR for your branch and I'll push a commit that supports this if you don't have the time to work on it. I don't want to give you more homework than you already have 😛

**Comment by novafacing** [comment]
I've implemented the `path_seprator` logic but it seems that recursive library loads may still be broken (impl @ https://github.com/novafacing/maat/commit/c11e166f40c82cb5a437cb92e037ee0865c64218).

When I run the same example I've been working on which has a callback every time the program maps change, I get:
```
**[Info] Adding library 'ld-linux-x86-64.so.2' to emulated filesystem at '/usr/lib/ld-linux-x86-64.so.2'
[Info] Adding library 'libc.so.6' to emulated filesystem at '/lib/x86_64-linux-gnu/libc.so.6'
[Info] Adding library 'libcgc.so' to emulated filesystem at '/AIS-Lite/lib_include/libcgc.so'
[Info] Adding library 'libm.so.6' to emulated filesystem at '/lib/x86_64-linux-gnu/libm.so.6'
[Info] Adding library 'libtiny-AES128-C.so' to emulated filesystem at '/AIS-Lite/lib_include/tiny-AES128-C/libtiny-AES128-C.so'
[Info] Loading binary 'AIS-Lite' from path '//AIS-Lite/AIS-Lite'**
Setting callback at 0x00000000205a0a`
Initial maps {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'Stack': [(8796090925056, 8796093022207)]}`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'Stack': [(8796090925056, 8796093022207)]}`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'map_anon': [(67108864, 67117055)], 'Stack': [(8796090925056, 8796093022207)]}`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'map_anon': [(67108864, 67117055)], '/AIS-Lite/lib_include/libcgc.so': [(67117056, 67149823)], 'Stack': [(8796090925056, 8796093022207)]}`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'map_anon': [(67108864, 67117055)], '/AIS-Lite/lib_include/libcgc.so': [(67117056, 67129343), (67129344, 67141631), (67141632, 67149823)], 'Stack': [(8796090925056, 8796093022207)]}`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'map_anon': [(67108864, 67117055)], '/AIS-Lite/lib_include/libcgc.so': [(67117056, 67129343), (67129344, 67141631), (67141632, 67145727), (67145728, 67149823)], 'Stack': [(8796090925056, 8796093022207)]}`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by AIS-Lite)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)`
Maps changed: {'ld-linux-x86-64.so.2': [(4096, 8191), (8192, 151551), (151552, 184319), (188416, 200703)], 'Interp.': [(200704, 4395007)], 'Heap': [(4395008, 6324223)], 'map_anon': [(67108864, 67117055), (67149824, 67158015)], '/AIS-Lite/lib_include/libcgc.so': [(67117056, 67129343), (67129344, 67141631), (67141632, 67145727), (67145728, 67149823)], 'Stack': [(8796090925056, 8796093022207)]}`
AIS-Lite: symbol lookup error: AIS-Lite: undefined symbol: __libc_start_main, version GLIBC_2.2.5`
[Error] Exception in CALLOTHER handler: SYSCALL: EnvEmulator: syscall '231' not supported for emulation`
[Error] Unexpected error when processing IR instruction, aborting...`
```

**Comment by Boyan-MILANOV** [comment]
Quick update: I've looked into the issue a bit more and I found out that when executing the loader with Maat it refuses to map some of the libraries (it opens them, reads some info in the headers, and then closes the file without `mmap`ing it), which causes the `undefined symbol: __libc_start_main, version GLIBC_2.2.5` error later. 

I just generated execution traces for both cases, running the loader in Maat and running the loader on a Linux computer, both with the exact same loader and libraries. I have found the exact branch where execution diverges, I'll try to find the root cause starting from there.

**Comment by novafacing** [comment]
@Boyan-MILANOV great! I'll stay posted on that and re-test this once that works :)

**Comment by Boyan-MILANOV** [comment]
It took me forever but I finally found what causes the loader to abort loading libraries when executing in Maat.

**tl;dr**: our `fstat` syscall callback was returning the same inode number for all libraries loaded in the emulated filesystem. This caused the loader to think that all libraries were the same file, and thus load only the first library and ignore the next ones (including `libc.so`), leading to the `undefined symbol: __libc_start_main, version GLIBC_2.2.5` error later on.

**note**: Even after the fix I still see the `AIS-Lite: /AIS-Lite/lib_include/libcgc.so: no version information available (required by /AIS-L)` messages, but the program seems to run fine so I don't think this is an issue. 

@novafacing: I have a local branch with the fix for this issue and a couple other small improvements/fixes that I made during the debugging process. I'll clean it up a bit and try to push it later today so that you can test it.

<hr>

#### More details for reference

When mapping a new library, `ld.so` checks whether the file has already been loaded. If yes then it doesn't map it again and just uses the already mapped file. The code doing this can be found in [dl-load.c](https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/elf/dl-load.c#L891-L905)

```
  /* Look again to see if the real name matched another already loaded.  */
  for (l = GL(dl_ns)[nsid]._ns_loaded; l != NULL; l = l->l_next)
    if (!l->l_removed && _dl_file_id_match_p (&l->l_file_id, &id))
      {
	/* The object is already loaded.
	   Just bump its reference count and return it.  */
	__close_nocancel (fd);

	/* If the name is not in the list of names for this object add
	   it.  */
	free (realname);
	add_name_to_object (l, name);

	return l;
      }
```

To check whether the file was already loaded, it uses `_dl_file_id_match_p()` which compares files using their device ID and inode number information, as returned by the `stat()` system call.

```
_dl_file_id_match_p (const struct r_file_id *a, const struct r_file_id *b)
{
return a->dev == b->dev && a->ino == b->ino;
}
```

Some info on how I debugged this that might be useful if similar bugs show up:
- Trying to diff a Maat trace vs a native trace unfortunately doesn't work so great, because there will always be environment differences that will cause the traces to diverge because of a different `envp`, auxiliary vector, etc
- I started by making a syscall trace in Maat and noticed that the loader was opening `libc.so`, calling the `stat` syscall on it, and then closing it, without doing any `mmap`. That tells us something is likely going wrong there.
- I made a second trace in Maat (this time tracing all instructions), between the `stat` syscall and the moment the loader closes `libc.so`. Then I opened `ld.so` in a disassembler to see what part of the loader that corresponds to.
- Unfortunately `ld.so` contains almost no symbols, and it was just too hard to reverse what was going on manually. Solution: run `AIS-Lite` in `gdb` and load the `libc6-dbg` debug binary for `ld-linux`. Put a breakpoint on the code corresponding to the Maat trace, run, and then thanks to the debug info `gdb` can tell us that we are in the `_dl_map_object_from_fd` function
    - `gdb` can't load debug info automatically for the loader and I had to follow [this workaround](https://bugs.launchpad.net/ubuntu/+source/eglibc/+bug/793292):
    
      >  A workaround is the following series of commands in gdb:
      >  (gdb) info share ld-linux
      >  Note down the address in the "from" column and substitute it for ADDRESS in the next command.
      >  (gdb) add-symbol-file /usr/lib/debug/lib/x86_64-linux-gnu/ld-2.13.so ADDRESS
- From there, with both the libc source and the asm trace, finding the cause of the issue is pretty straightforward

**Comment by novafacing** [comment]
Awesome! Do we still want to merge in the `virtual_fs` changes? I can see it still potentially being useful for `dlopen` library loads and some other edge cases, but I'll leave it up to you.

**Comment by Boyan-MILANOV** [comment]
Well the `virtual_fs` changes give more flexibility (and if it's not specified we still default to loading files at standard locations anyway) so I think it's worth merging them :) 

Would you mind rebasing on master and then open a PR for it?
